### ISSUE TITLE
FlatSharp.Compiler doesn't produce valid code

### ISSUE THREAD

**Issue by HotelCalifornia** [jamescourtney/FlatSharp#84]
I want to preface this by saying I'm nearly certain I've done something wrong---my normal wheelhouse is C++, not C#.

I have a Winforms project targeting .NET Framework 4.7.2 in Visual Studio Professional 2019 and I've installed FlatSharp.Runtime + FlatSharp.Compiler, both at version 4.1.0 with NuGet.

Building the project results in the following errors:

```
Rebuild started...
1>------ Rebuild All started: Project: SYCLCommsTester, Configuration: Debug Any CPU ------
1>  Compiling 
1>  C:\Users\<...>\source\repos\SYCLCommsTester\packages\FlatSharp.Compiler.4.1.0\tools\net47\FlatSharp.Compiler.exe C:\Users\<...>\source\repos\SYCLCommsTester\SYCLCommsTester\Flatbuffers\SYCL.fbs
1>C:\Users\<...>\source\repos\SYCLCommsTester\SYCLCommsTester\obj\Debug\SYCL.fbs.generated.cs(67,18,67,23): error CS1001: Identifier expected
1>C:\Users\<...>\source\repos\SYCLCommsTester\SYCLCommsTester\obj\Debug\SYCL.fbs.generated.cs(67,18,67,23): error CS1002: ; expected
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========
```

For comparison, I tried using `flatc` directly, which appeared to produce valid code (although I don't currently have the FlatBuffers C# code installed in the project to check).

Further, I'm pretty sure the schema is valid as the same one works without incident in a C++ project.

Intellisense seems to think there might be a missing using directive or assembly reference, and sure enough, FlatSharp.Compiler isn't listed in visual studio's list of references for the project... though given my lack of experience with C# I can't say whether it's supposed to be there or not.

As an aside, it wasn't really clear to me from the readme which packages I'm meant to install, so while writing this issue I ended up installing FlatSharp (also v 4.1.0) itself as well to see if that fixed things (it didn't) 😅 

Attached below are relevant files.

<details>
<summary>csproj file</summary>

```xml
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
  <PropertyGroup>
    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
    <ProjectGuid>{13F04FF5-53E7-4D70-A600-3CE96960CCEF}</ProjectGuid>
    <OutputType>WinExe</OutputType>
    <RootNamespace>SYCLCommsTester</RootNamespace>
    <AssemblyName>SYCLCommsTester</AssemblyName>
    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
    <FileAlignment>512</FileAlignment>
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <Deterministic>true</Deterministic>
    <NuGetPackageImportStamp>
    </NuGetPackageImportStamp>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <PlatformTarget>AnyCPU</PlatformTarget>
    <DebugSymbols>true</DebugSymbols>
    <DebugType>full</DebugType>
    <Optimize>false</Optimize>
    <OutputPath>bin\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <PlatformTarget>AnyCPU</PlatformTarget>
    <DebugType>pdbonly</DebugType>
    <Optimize>true</Optimize>
    <OutputPath>bin\Release\</OutputPath>
    <DefineConstants>TRACE</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
  </PropertyGroup>
  <ImportGroup Label="ExtensionTargets">
    <Import Project="flatbuffers.targets" />
    <Import Project="..\packages\FlatSharp.Compiler.4.1.0\build\net47\FlatSharp.Compiler.targets" Condition="Exists('..\packages\FlatSharp.Compiler.4.1.0\build\net47\FlatSharp.Compiler.targets')" />
  </ImportGroup>
  <ItemGroup>
    <Reference Include="FlatSharp.Runtime, Version=4.1.0.0, Culture=neutral, PublicKeyToken=06041f83a2658c5b, processorArchitecture=MSIL">
      <HintPath>..\packages\FlatSharp.Runtime.4.1.0\lib\netstandard2.0\FlatSharp.Runtime.dll</HintPath>
    </Reference>
    <Reference Include="System">
      <HintPath>..\..\..\..\..\..\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.7.2\System.dll</HintPath>
    </Reference>
    <Reference Include="System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Buffers.4.4.0\lib\netstandard2.0\System.Buffers.dll</HintPath>
    </Reference>
    <Reference Include="System.Core" />
    <Reference Include="System.Memory, Version=4.0.1.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Memory.4.5.3\lib\netstandard2.0\System.Memory.dll</HintPath>
    </Reference>
    <Reference Include="System.Numerics" />
    <Reference Include="System.Numerics.Vectors, Version=4.1.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Numerics.Vectors.4.4.0\lib\net46\System.Numerics.Vectors.dll</HintPath>
    </Reference>
    <Reference Include="System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Runtime.CompilerServices.Unsafe.4.5.2\lib\netstandard2.0\System.Runtime.CompilerServices.Unsafe.dll</HintPath>
    </Reference>
    <Reference Include="System.Xml.Linq" />
    <Reference Include="System.Data.DataSetExtensions" />
    <Reference Include="Microsoft.CSharp" />
    <Reference Include="System.Data" />
    <Reference Include="System.Deployment" />
    <Reference Include="System.Drawing" />
    <Reference Include="System.Net.Http" />
    <Reference Include="System.Windows.Forms" />
    <Reference Include="System.Xml" />
  </ItemGroup>
  <ItemGroup>
    <FlatSharpSchema Include="Flatbuffers\SYCL.fbs" />
    <Compile Include="Form1.cs">
      <SubType>Form</SubType>
    </Compile>
    <Compile Include="Form1.Designer.cs">
      <DependentUpon>Form1.cs</DependentUpon>
    </Compile>
    <Compile Include="Program.cs" />
    <Compile Include="Properties\AssemblyInfo.cs" />
    <EmbeddedResource Include="Form1.resx">
      <DependentUpon>Form1.cs</DependentUpon>
    </EmbeddedResource>
    <EmbeddedResource Include="Properties\Resources.resx">
      <Generator>ResXFileCodeGenerator</Generator>
      <LastGenOutput>Resources.Designer.cs</LastGenOutput>
      <SubType>Designer</SubType>
    </EmbeddedResource>
    <Compile Include="Properties\Resources.Designer.cs">
      <AutoGen>True</AutoGen>
      <DependentUpon>Resources.resx</DependentUpon>
    </Compile>
    <None Include="packages.config" />
    <None Include="Properties\Settings.settings">
      <Generator>SettingsSingleFileGenerator</Generator>
      <LastGenOutput>Settings.Designer.cs</LastGenOutput>
    </None>
    <Compile Include="Properties\Settings.Designer.cs">
      <AutoGen>True</AutoGen>
      <DependentUpon>Settings.settings</DependentUpon>
      <DesignTimeSharedInput>True</DesignTimeSharedInput>
    </Compile>
  </ItemGroup>
  <ItemGroup>
    <None Include="App.config" />
  </ItemGroup>
  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
  <PropertyGroup>
    <PreBuildEvent>
    </PreBuildEvent>
  </PropertyGroup>
  <Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">
    <PropertyGroup>
      <ErrorText>This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
    </PropertyGroup>
    <Error Condition="!Exists('..\packages\FlatSharp.Compiler.4.1.0\build\net47\FlatSharp.Compiler.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\FlatSharp.Compiler.4.1.0\build\net47\FlatSharp.Compiler.targets'))" />
  </Target>
</Project>
```

</details>

<details>
<summary>.fbs file</summary>

```fbs
namespace SYCL;

enum MessageType : byte { None = -2, Nack, Ack }

enum EventType : byte {
	evInvalid = -1,
	evReady,
	evLockCovers,
	evCanUnlock,
	evCanUnlockReceived,
	evEnableUnlockCovers,
	evButtonUnlock,
	evPrepareUnlockComplete,
	evUnlockAllowed,
	evUnlockAllowedReceived,
	evCoverOpen,
	evCoverClosed,
	evTimer,
	evAbort,
	evPeerAbort,
	evPeerAbortReceived,
	evPeerPresent,
	evPeerCoversUnlocked,
	evPeerCoversUnlockedReceived,
	evPeerLockCovers,
	evPeerLockCoversReceived,
	evPeerCoversLocked,
	evPeerCoversLockedReceived,
	evResumeComplete
}

table Message {
	type:MessageType = None;
	event:EventType = evInvalid;
}

root_type Message;
```

</details>

<details>
<summary>full C# source generated by FlatSharp.Compiler 4.1.0</summary>

```cs

//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by the FlatSharp FBS to C# compiler (source hash: 4.1.0.rpjvg+iPJ95nZOpRm7z5r/ZNW2oz0cV66tShV6dLRQE=)
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using FlatSharp;
using FlatSharp.Attributes;
namespace SYCL
{
    [FlatBufferEnum(typeof(System.SByte))]
    [System.Runtime.CompilerServices.CompilerGenerated]
    public enum MessageType : System.SByte
    {
        None = (sbyte)(-2),
        Nack = (sbyte)(-1),
        Ack = (sbyte)(0),
    }
    
    [FlatBufferEnum(typeof(System.SByte))]
    [System.Runtime.CompilerServices.CompilerGenerated]
    public enum EventType : System.SByte
    {
        evInvalid = (sbyte)(-1),
        evReady = (sbyte)(0),
        evLockCovers = (sbyte)(1),
        evCanUnlock = (sbyte)(2),
        evCanUnlockReceived = (sbyte)(3),
        evEnableUnlockCovers = (sbyte)(4),
        evButtonUnlock = (sbyte)(5),
        evPrepareUnlockComplete = (sbyte)(6),
        evUnlockAllowed = (sbyte)(7),
        evUnlockAllowedReceived = (sbyte)(8),
        evCoverOpen = (sbyte)(9),
        evCoverClosed = (sbyte)(10),
        evTimer = (sbyte)(11),
        evAbort = (sbyte)(12),
        evPeerAbort = (sbyte)(13),
        evPeerAbortReceived = (sbyte)(14),
        evPeerPresent = (sbyte)(15),
        evPeerCoversUnlocked = (sbyte)(16),
        evPeerCoversUnlockedReceived = (sbyte)(17),
        evPeerLockCovers = (sbyte)(18),
        evPeerLockCoversReceived = (sbyte)(19),
        evPeerCoversLocked = (sbyte)(20),
        evPeerCoversLockedReceived = (sbyte)(21),
        evResumeComplete = (sbyte)(22),
    }
    
    [FlatBufferTable]
    [System.Runtime.CompilerServices.CompilerGenerated]
    public partial class Message : object
    {
        partial void OnInitialized();
         public Message() { this.OnInitialized(); }
        public Message(Message source)
        {
            this.type = source.type;
            this.event = source.event;
            this.OnInitialized();
        }
        
        [FlatBufferItem(0, DefaultValue = global::SYCL.MessageType.None)]
        public virtual global::SYCL.MessageType type { get;  set; } = global::SYCL.MessageType.None;
        [FlatBufferItem(1, DefaultValue = global::SYCL.EventType.evInvalid)]
        public virtual global::SYCL.EventType event { get;  set; } = global::SYCL.EventType.evInvalid;
    }
}
```

</details>

<details>
<summary>full C# file generated by flatc.exe version 1.12.0</summary>

```cs
// <auto-generated>
//  automatically generated by the FlatBuffers compiler, do not modify
// </auto-generated>

namespace SYCL
{

using global::System;
using global::System.Collections.Generic;
using global::FlatBuffers;

public enum MessageType : sbyte
{
  None = -2,
  Nack = -1,
  Ack = 0,
};

public enum EventType : sbyte
{
  evInvalid = -1,
  evReady = 0,
  evLockCovers = 1,
  evCanUnlock = 2,
  evCanUnlockReceived = 3,
  evEnableUnlockCovers = 4,
  evButtonUnlock = 5,
  evPrepareUnlockComplete = 6,
  evUnlockAllowed = 7,
  evUnlockAllowedReceived = 8,
  evCoverOpen = 9,
  evCoverClosed = 10,
  evTimer = 11,
  evAbort = 12,
  evPeerAbort = 13,
  evPeerAbortReceived = 14,
  evPeerPresent = 15,
  evPeerCoversUnlocked = 16,
  evPeerCoversUnlockedReceived = 17,
  evPeerLockCovers = 18,
  evPeerLockCoversReceived = 19,
  evPeerCoversLocked = 20,
  evPeerCoversLockedReceived = 21,
  evResumeComplete = 22,
};

public struct Message : IFlatbufferObject
{
  private Table __p;
  public ByteBuffer ByteBuffer { get { return __p.bb; } }
  public static void ValidateVersion() { FlatBufferConstants.FLATBUFFERS_1_12_0(); }
  public static Message GetRootAsMessage(ByteBuffer _bb) { return GetRootAsMessage(_bb, new Message()); }
  public static Message GetRootAsMessage(ByteBuffer _bb, Message obj) { return (obj.__assign(_bb.GetInt(_bb.Position) + _bb.Position, _bb)); }
  public void __init(int _i, ByteBuffer _bb) { __p = new Table(_i, _bb); }
  public Message __assign(int _i, ByteBuffer _bb) { __init(_i, _bb); return this; }

  public SYCL.MessageType Type { get { int o = __p.__offset(4); return o != 0 ? (SYCL.MessageType)__p.bb.GetSbyte(o + __p.bb_pos) : SYCL.MessageType.None; } }
  public SYCL.EventType Event { get { int o = __p.__offset(6); return o != 0 ? (SYCL.EventType)__p.bb.GetSbyte(o + __p.bb_pos) : SYCL.EventType.evInvalid; } }

  public static Offset<SYCL.Message> CreateMessage(FlatBufferBuilder builder,
      SYCL.MessageType type = SYCL.MessageType.None,
      SYCL.EventType event = SYCL.EventType.evInvalid) {
    builder.StartTable(2);
    Message.AddEvent(builder, event);
    Message.AddType(builder, type);
    return Message.EndMessage(builder);
  }

  public static void StartMessage(FlatBufferBuilder builder) { builder.StartTable(2); }
  public static void AddType(FlatBufferBuilder builder, SYCL.MessageType type) { builder.AddSbyte(0, (sbyte)type, -2); }
  public static void AddEvent(FlatBufferBuilder builder, SYCL.EventType event) { builder.AddSbyte(1, (sbyte)event, -1); }
  public static Offset<SYCL.Message> EndMessage(FlatBufferBuilder builder) {
    int o = builder.EndTable();
    return new Offset<SYCL.Message>(o);
  }
  public static void FinishMessageBuffer(FlatBufferBuilder builder, Offset<SYCL.Message> offset) { builder.Finish(offset.Value); }
  public static void FinishSizePrefixedMessageBuffer(FlatBufferBuilder builder, Offset<SYCL.Message> offset) { builder.FinishSizePrefixed(offset.Value); }
};


}
```

**Comment by HotelCalifornia** [comment]
worse comes to worst I can go back to using flatc directly, but I landed on this project after a couple hours of banging my head against MSBuild so I'd prefer to stick with this if at all possible

**Comment by jamescourtney** [comment]
I had to throw it into VS and look at syntax highlighting. Turns out 'event' is a keyword, and that's making the compiler grumpy.

I'll mark this as a bug. In the meantime, I you can just use big-'E' event or change to 'eventType'. Thanks for reaching out!

**Comment by HotelCalifornia** [comment]
ah-ha, i should have realised. thanks for looking into this!

**Comment by jamescourtney** [comment]
This is resolved in #114. Flatsharp compiler now returns the general context around the error it encountered. This will be shipped in version 5 once it is released.
